gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -0,0 +1,861 @@
+  // eigenvectors corresponding to the eigenvalue 0.  In the case  that the
+  // relaxation is exact, this is the *minimum* eigenvalue, and therefore the
+  // rows of S are exactly the eigenvectors that we're looking for.  On the
+  // other hand, if the relaxation is *not* exact, then S - Lambda(S) has at
+  // least one strictly negative eigenvalue, and the rows of S are *unstable
+  // fixed points* for the Lanczos iterations.  Thus, we will take a slightly
+  // "fuzzed" version of the first row of S as an initialization for the
+  // Lanczos iterations; this allows for rapid convergence in the case that
+  // the relaxation is exact (since are starting close to a solution), while
+  // simultaneously allowing the iterations to escape from this fixed point in
+  // the case that the relaxation is not exact.
+  Vector v0 = S.row(0).transpose();
+  Vector perturbation(v0.size());
+  perturbation.setRandom();
+  perturbation.normalize();
+  Vector xinit = v0 + (.03 * v0.norm()) * perturbation; // Perturb v0 by ~3%
+
+  // Use this to initialize the eigensolver
+  minEigenValueSolver.init(xinit.data());
+
+  // Now determine the relative precision required in the Lanczos method in
+  // order to be able to estimate the smallest eigenvalue within an *absolute*
+  // tolerance of 'minEigenvalueNonnegativityTolerance'
+  const int minConverged = minEigenValueSolver.compute(
+      maxIterations, minEigenvalueNonnegativityTolerance / lmEigenValue,
+      Spectra::SELECT_EIGENVALUE::LARGEST_MAGN);
+
+  if (minConverged != 1)
+    return false;
+
+  *minEigenValue = minEigenValueSolver.eigenvalues()(0) + 2 * lmEigenValue;
+  if (minEigenVector) {
+    *minEigenVector = minEigenValueSolver.eigenvectors(1);
+    minEigenVector->normalize(); // Ensure that this is a unit vector
+  }
+  if (numIterations)
+    *numIterations = minEigenValueSolver.num_iterations();
+  return true;
+}
+
+/* ************************************************************************* */
+template <size_t d> Sparse ShonanAveraging<d>::computeA(const Matrix &S) const {
+  auto Lambda = computeLambda(S);
+  return Lambda - Q_;
+}
+
+/* ************************************************************************* */
+template <size_t d>
+double ShonanAveraging<d>::computeMinEigenValue(const Values &values,
+                                                Vector *minEigenVector) const {
+  assert(values.size() == nrUnknowns());
+  const Matrix S = StiefelElementMatrix<d>(values);
+  auto A = computeA(S);
+#ifdef SLOW_EIGEN_COMPUTATION
+  Eigen::EigenSolver<Matrix> solver(Matrix(A), false);
+  auto lambdas = solver.eigenvalues();
+  double minEigenValue = lambdas(0).real();
+  for (size_t i = 1; i < lambdas.size(); i++) {
+    minEigenValue = min(lambdas(i).real(), minEigenValue);
+  }
+#else
+  double minEigenValue;
+  bool success = SparseMinimumEigenValue(A, S, &minEigenValue, minEigenVector);
+  if (!success) {
+    throw std::runtime_error(
```

Comment:
Probably should fall back to the slow version instead of bailing out?

Reply:
That is *extremely* slow, and for large problems will sink your machine :-) I removed the slow code and moved it to the tests.